pystiche/papers/sanakoyeu_et_al_2018/loss.py

```diff
@@ -0,0 +1,237 @@
+    real_predictions = discriminator(real_image)
+    discr_loss = loss(fake_predictions, real=False, scale_weight=scale_weight) + loss(
+        real_predictions, real=True, scale_weight=scale_weight
+    )
+    discr_acc = (
+        acc(fake_predictions, real=False) + acc(real_predictions, real=True)
+    ) / 2
+    return discr_loss, discr_acc
+
+
+class DiscriminatorLoss(nn.Module):
+    def __init__(self, discriminator: SanakoyeuEtAl2018Discriminator) -> None:
+        super().__init__()
+        self.discriminator = discriminator
+        self.acc = 0.0
+
+    def get_current_acc(self):
+        return self.acc
+
+    def forward(
+        self, fake_image: torch.Tensor, real_image: torch.Tensor
+    ) -> torch.Tensor:
+        loss, self.acc = discriminator_loss(fake_image, real_image, self.discriminator)
+        return loss
+
+
+class SanakoyeuEtAl2018FeatureOperator(MSEEncodingOperator):
+    def __init__(
+        self, encoder: Encoder, score_weight: float = 1.0, impl_params: bool = True
+    ):
+        super().__init__(encoder, score_weight=score_weight)
+        self.impl_params = impl_params
+
+    def update_encoder(self, encoder: Encoder):
+        self.encoder = encoder  # FIXME: necessary?
+
+    def calculate_score(
+        self, input_repr: torch.Tensor, target_repr: torch.Tensor, ctx: None
+    ) -> torch.Tensor:
+        if self.impl_params:
+            return torch.mean(torch.abs(input_repr - target_repr))
+        else:
+            return super().calculate_score(input_repr, target_repr, ctx)
+
+
+def sanakoyeu_et_al_2018_style_aware_content_loss(
+    encoder: Optional[pystiche.SequentialModule],
+    impl_params: bool = True,
+    score_weight=None,
+) -> SanakoyeuEtAl2018FeatureOperator:
+    if score_weight is None:
+        if impl_params:
+            score_weight = 1e2
+        else:
+            score_weight = 1e0
+
+    return SanakoyeuEtAl2018FeatureOperator(
+        encoder, score_weight=score_weight, impl_params=impl_params
+    )
+
+
+class SanakoyeuEtAl2018TransformerOperator(MSEEncodingOperator):
+    def update_transformer(self, transformer: Encoder):
+        self.encoder = transformer  # FIXME: necessary?
+
+
+def sanakoyeu_et_al_2018_transformer_loss(
+    impl_params: bool = True,
+    transformer: Optional[SanakoyeuEtAl2018TransformerBlock] = None,
+    score_weight=None,
+) -> SanakoyeuEtAl2018TransformerOperator:
+    if score_weight is None:
+        if impl_params:
+            score_weight = 1e2
+        else:
+            score_weight = 1e0
+
+    if transformer is None:
+        transformer = SanakoyeuEtAl2018TransformerBlock()
+
+    return SanakoyeuEtAl2018TransformerOperator(transformer, score_weight=score_weight)
+
+
+class SanakoyeuEtAl2018GeneratorLoss(MultiOperatorLoss):
+    def __init__(
+        self,
+        style_aware_content_loss: SanakoyeuEtAl2018FeatureOperator,
+        transformer_loss: SanakoyeuEtAl2018TransformerOperator,
+        discr_loss: DiscriminatorOperator,
```

Comment:
You mean like a `class SanakoyeuEtAl2018ContentOperator(MultiOperatorLoss)` to group them together?

Reply:
`MultiOperatorLoss` is the _last_ instance. If you want to simply group multiple operators you can use an [`OperatorContainer`](). Here that would be

```python
from collections import OrderedDict
from pystiche.ops import OperatorContainer

content_loss = OperatorContainer(
    OrderedDict(
        (
            ("style_aware_content_loss", style_aware_content_loss),
            ("tranformer_loss", tranformer_loss),
        )
    )
)
```

Reply:
If I understand the `class` correctly, I need to use a modified version of this `OperatorContainer` to get a function `set_target_image`. Am I right?
The function is necessary because the operators based on `ComparisonOperator` in which the `content_image` is compared with the `stylized_image`.

Reply:
Yeah, currently this is not natively supported. I'm working on something like this, but it is not finished yet. For now you can create a `ContentOperatorContainer` class and implement a `set_target` method.